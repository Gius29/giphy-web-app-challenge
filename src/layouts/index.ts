// Each constant represent a possible layout that the app may have.
export enum LayoutIDs {
  ONE_COLUMN = "ONE_COLUMN",
  THREE_COLUMNS = "THREE_COLUMNS"
}
