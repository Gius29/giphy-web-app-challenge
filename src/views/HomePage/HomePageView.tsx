import React, { Component } from "react";
import { connect } from "react-redux";
import {
  searchQueryAction,
  deleteSearchAction,
  changeLayoutAction
} from "../../redux/actions";
import { AppState } from "../../redux/state/AppState";
import { HomePageState } from "../../redux/state/HomePageState";

import { InfiniteScroll } from "../../components/InfiniteScroll";
import { Card, CardModel } from "../../components/Card";
import { SearchBar } from "../../components/SearchBar";

import styles from "./HomePageView.module.scss";

interface DispatchProps {
  searchQueryAction: (limit: number, offset: number, query: string) => void;
  deleteSearchAction: () => void;
  changeLayoutAction: () => void;
}

type HomePageViewPops = HomePageState & DispatchProps;

export class HomePageView extends Component<HomePageViewPops> {
  readonly searchPlaceHolderMsg: string;

  constructor(props: HomePageViewPops) {
    super(props);
    this.searchPlaceHolderMsg = "Enter your query here";

    this.newQueryFromSearchBar = this.newQueryFromSearchBar.bind(this);
    this.nextPaginationAfterSearch = this.nextPaginationAfterSearch.bind(this);
  }

  newQueryFromSearchBar(query: string) {
    // Dispatch an action to delete previus results and reset the status
    this.props.deleteSearchAction();

    // Dispatch an action to fetch new results using the GIPHY API
    this.props.searchQueryAction(this.props.searchLimit, 0, query);
  }

  nextPaginationAfterSearch() {
    // if there is an ongoing pagination request we don't start a new one
    if (this.props.searchResultPending) {
      return;
    }
    this.props.searchQueryAction(
      this.props.searchLimit,
      this.props.searchOffset,
      this.props.query
    );
  }

  render() {
    return (
      <div className={styles.homePageContainer}>
        <h2> Giphy Search Challenge </h2>
        <SearchBar
          placeHolderMsg={this.searchPlaceHolderMsg}
          searchQuery={this.newQueryFromSearchBar}
          layoutID={this.props.layoutID}
          changeLayout={this.props.changeLayoutAction}
        />

        <InfiniteScroll
          items={this.props.cards}
          render={(card: CardModel) => <Card card={card} />}
          searchQuery={this.nextPaginationAfterSearch}
          layoutID={this.props.layoutID}
        />

        {this.props.searchResultPending && (
          <div className={styles.message}>Searching...</div>
        )}

        {!this.props.searchResultPending &&
          this.props.searchOffset === -1 && (
            <div className={styles.message}>End of the list</div>
          )}
      </div>
    );
  }
}

// [IMPROVMENT] We can use a selector library to simplify the syntax
// Ex. https://github.com/reduxjs/reselect
const mapStateToProps = (state: AppState) => {
  if (state.homePage) {
    const {
      cards,
      query,
      searchOffset,
      searchLimit,
      searchResultPending,
      layoutID
    } = state.homePage;
    return {
      cards,
      query,
      searchOffset,
      searchLimit,
      searchResultPending,
      layoutID
    };
  }

  return state;
};

const mapDispatchToProps: DispatchProps = {
  searchQueryAction,
  deleteSearchAction,
  changeLayoutAction
};

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(HomePageView);
