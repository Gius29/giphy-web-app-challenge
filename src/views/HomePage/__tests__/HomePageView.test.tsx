import React from "react";
import renderer from "react-test-renderer";
import { mount } from "enzyme";
import { LayoutIDs } from "../../../layouts";
import { HomePageView } from "../HomePageView";

describe("HomePageView", () => {
  it("should renders correctly with ONE_COLUMN layout and a pagination pending", () => {
    const props = {
      cards: [],
      query: "",
      searchOffset: 0,
      searchLimit: 10,
      searchResultPending: true,
      layoutID: LayoutIDs.ONE_COLUMN
    };
    const wrapper = renderer.create(<HomePageView {...props} />);

    expect(wrapper).toMatchSnapshot();
  });

  it("should renders correctly with ONE_COLUMN layout and end of the list", () => {
    const props = {
      cards: [],
      query: "",
      searchOffset: -1,
      searchLimit: 10,
      searchResultPending: false,
      layoutID: LayoutIDs.THREE_COLUMNS
    };
    const wrapper = renderer.create(<HomePageView {...props} />);

    expect(wrapper).toMatchSnapshot();
  });

  it("should provide InfiniteScroll with the ability to perform a searchQuery", () => {
    const searchQueryAction = jest.fn();
    const props = {
      cards: [],
      query: "query",
      searchOffset: 0,
      searchLimit: 10,
      searchResultPending: false,
      layoutID: LayoutIDs.THREE_COLUMNS,
      searchQueryAction
    };
    const wrapper = mount(<HomePageView {...props} />);

    // it simulates the user that scroll down and triggers the next pagination of the search
    wrapper
      .find("InfiniteScroll")
      .props()
      .searchQuery();

    wrapper.unmount();
    expect(searchQueryAction).toHaveBeenCalledTimes(1);
    expect(searchQueryAction).toHaveBeenCalledWith(10, 0, "query");
  });

  it("should block a search query from InfiniteScroll if searchResultPending=true", () => {
    const searchQueryAction = jest.fn();
    const props = {
      cards: [],
      query: "query",
      searchOffset: 10,
      searchLimit: 5,
      searchResultPending: true,
      layoutID: LayoutIDs.THREE_COLUMNS,
      searchQueryAction
    };
    const wrapper = mount(<HomePageView {...props} />);

    // it simulates the user that scroll down and triggers the next pagination of the search
    wrapper
      .find("InfiniteScroll")
      .props()
      .searchQuery();

    wrapper.unmount();
    expect(searchQueryAction).toHaveBeenCalledTimes(0);
  });

  it("should provide SearchBar with the ability to perform a searchQuery", () => {
    const searchQueryAction = jest.fn();
    const deleteSearchAction = jest.fn();
    const props = {
      cards: [],
      query: "query",
      searchOffset: 10,
      searchLimit: 5,
      searchResultPending: false,
      layoutID: LayoutIDs.THREE_COLUMNS,
      searchQueryAction,
      deleteSearchAction
    };
    const wrapper = mount(<HomePageView {...props} />);

    // it simulates a new search done by the user
    wrapper
      .find("SearchBar")
      .props()
      .searchQuery("newSearchString");

    wrapper.unmount();
    expect(deleteSearchAction).toHaveBeenCalledTimes(1);
    expect(searchQueryAction).toHaveBeenCalledTimes(1);
    expect(searchQueryAction).toHaveBeenCalledWith(5, 0, "newSearchString");
  });
});
