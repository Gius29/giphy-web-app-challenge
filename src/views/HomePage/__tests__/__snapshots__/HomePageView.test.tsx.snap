// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HomePageView should renders correctly with ONE_COLUMN layout and a pagination pending 1`] = `
<div
  className="homePageContainer"
>
  <h2>
     Giphy Search Challenge 
  </h2>
  <form
    className="searchBar"
    onSubmit={[Function]}
  >
    <input
      onChange={[Function]}
      placeholder="Enter your query here"
      value=""
    />
    <button
      type="submit"
    >
      Search
    </button>
    <button
      type="button"
    >
      3 columns
    </button>
  </form>
  <ul />
  <div
    className="message"
  >
    Searching...
  </div>
</div>
`;

exports[`HomePageView should renders correctly with ONE_COLUMN layout and end of the list 1`] = `
<div
  className="homePageContainer"
>
  <h2>
     Giphy Search Challenge 
  </h2>
  <form
    className="searchBar"
    onSubmit={[Function]}
  >
    <input
      onChange={[Function]}
      placeholder="Enter your query here"
      value=""
    />
    <button
      type="submit"
    >
      Search
    </button>
    <button
      type="button"
    >
      1 column
    </button>
  </form>
  <ul />
  <div
    className="message"
  >
    End of the list
  </div>
</div>
`;
